lib/mongo/server/connection_pool.rb

```diff
@@ -513,40 +530,102 @@ def close_idle_sockets
       end
 
       # Create and add connections to the pool until the
-      # pool size is at least min_size.
+      # pool size is at least min_size. Terminates if two connections
+      # encounter errors (in a row) while authenticating.
       #
       # Used by the pool populator background thread.
       #
+      # @return [ true | false ] True if populate ran to completion, false if it
+      # terminated because of errors encountered while connecting connections.
+      #
       # @api private
       def populate
-        return if closed?
+        raise_if_closed!
 
         catch(:done) do
+          retried = false
+
           loop do
+            connection = nil
             @lock.synchronize do
               if !closed? && unsynchronized_size < min_size
-                @available_connections << create_connection
+                connection = create_connection
+                @pending_connections << connection
               else
                 throw(:done)
               end
             end
+
+            begin
+              connect_connection(connection)
+            rescue Exception => e
+              @lock.synchronize do
+                @pending_connections.delete(connection)
+              end
+
+              if retried
+                return false
```

Comment:
This path does not signal the available semaphore. Do I have the following scenario correct:

- Pool is at max_size-1
- Populator starts to create another connection
- A thread wants a connection, can't create it in flow because pool is now at max_size, waits for the semaphore
- Populator fails with an error on the connection, but does not signal the semaphore
- Thread continues waiting although it should either abort with an error (current behavior we are going for as I understand it) or attempt to create the connection in flow (possible future behavior)
- Eventually the thread times out and raises the timeout exception

Reply:
Yes, this is a good point. If the semaphore is signaled after the retry fails, the thread will wake up and attempt to create the connection in flow, and error if that fails. I think this is preferred.